### PR TITLE
Add neighbor discovery behavior docs to kubeproxy-free.

### PR DESCRIPTION
**Pull request by bjhaid**
xref: https://github.com/cilium/cilium/pull/16974

Signed-off-by: Ayodele Abejide <abejideayodele@gmail.com>

### CODE REVIEW

File: Documentation/gettingstarted/kubeproxy-free.rst [hunk 1]
```diff
@@ -1222,6 +1222,15 @@ working, take a look at `this KEP
     free mode, make sure that default Kubernetes services like ``kube-dns`` and ``kubernetes``
     have the required label value.
 
+Neighbor Discovery
+******************
+
+When kube-proxy replacement is enabled Cilium does L2 neighbor discovery of nodes in the cluster, in
+some setups Cilium may leave stale entries behind in the neighbor table causing ethernet frames
+between some nodes to drop. To prevent Cilium from performing this neighbor discovery and instead
+rely on the Linux kernel to discover hosts on the same L2 network you can pass the
```
Comment (brb) [inline Documentation/gettingstarted/kubeproxy-free.rst]: Maybe you could extend the sentence saying that relying on the Linux kernel might cause some packets to be dropped, e.g., a NodePort request can be dropped on an intermediate node (i.e., the one which received and is going to forward to a destination node which runs the selected service endpoint) if there is no L2 neigh entry in the kernel, as BPF cannot drive the ARP resolution.

**Review by brb — CHANGES_REQUESTED** [review]
Thanks! Looks good, just one comment.

File: Documentation/gettingstarted/kubeproxy-free.rst (hunk 1 above)
Comment (bjhaid) [reply Documentation/gettingstarted/kubeproxy-free.rst]: thanks!

Comment (bjhaid) [reply Documentation/gettingstarted/kubeproxy-free.rst]: updated in https://github.com/cilium/cilium/commit/8b0776e755924acab39dc4b2c28bdbec792d7c12. PTAL

File: Documentation/gettingstarted/kubeproxy-free.rst [hunk 2]
```diff
@@ -1222,6 +1222,20 @@ working, take a look at `this KEP
...
+`--enable-l2-neigh-discovery` flag to the cilium-agent, setting it to false. Relying on the Linux
+Kernel might cause some packets to be dropped, e.g., a NodePort request can be dropped on an
+intermediate node (i.e., the one which received and is going to forward to a destination node which
+runs the selected service endpoint) if there is no L2 neigh entry in the kernel(due to the entry
```
Comment (brb) [inline Documentation/gettingstarted/kubeproxy-free.rst]: Nit: space after the kernel.

File: Documentation/gettingstarted/kubeproxy-free.rst [hunk 3]
```diff
@@ -1222,6 +1222,20 @@ working, take a look at `this KEP
     free mode, make sure that default Kubernetes services like ``kube-dns`` and ``kubernetes``
     have the required label value.
 
+Neighbor Discovery
+******************
+
+When kube-proxy replacement is enabled Cilium does L2 neighbor discovery of nodes in the cluster, in
```
Comment (brb) [inline Documentation/gettingstarted/kubeproxy-free.rst]: Let's make the second part as a separate sentence. Also, the second sentence:
> In some rare cases Cilium may leave stale entries behind in the neighbor table causing packets between some nodes to be dropped.

File: Documentation/gettingstarted/kubeproxy-free.rst [hunk 4]
```diff
@@ -1222,6 +1222,20 @@ working, take a look at `this KEP
...
+some setups Cilium may leave stale entries behind in the neighbor table causing ethernet frames
+between some nodes to drop. To prevent Cilium from performing this neighbor discovery and instead
```
Comment (brb) [inline Documentation/gettingstarted/kubeproxy-free.rst]: s/this neighbor discovery/the neighbor discovery/

File: Documentation/gettingstarted/kubeproxy-free.rst [hunk 5]
```diff
@@ -1222,6 +1222,20 @@ working, take a look at `this KEP
...
+rely on the Linux kernel to discover hosts on the same L2 network you can pass the
+`--enable-l2-neigh-discovery` flag to the cilium-agent, setting it to false. Relying on the Linux
+Kernel might cause some packets to be dropped, e.g., a NodePort request can be dropped on an
+intermediate node (i.e., the one which received and is going to forward to a destination node which
+runs the selected service endpoint) if there is no L2 neigh entry in the kernel(due to the entry
+being garbage collected or that ARP resolution has not been done by the kernel), as BPF cannot drive
+the ARP resolution.
```
Comment (brb) [inline Documentation/gettingstarted/kubeproxy-free.rst]: The sentence is very long, so let's break it down into two. The second:
> This is because Cilium does not drive the ARP resolution from the BPF programs.

**Review by brb — CHANGES_REQUESTED** [review]
A few more nits.

File: Documentation/gettingstarted/kubeproxy-free.rst (hunk 5 above)
Comment (bjhaid) [reply Documentation/gettingstarted/kubeproxy-free.rst]: Addressed in https://github.com/cilium/cilium/commit/262176f973869ede633241f226d038cf60a81de4

File: Documentation/gettingstarted/kubeproxy-free.rst (hunk 4 above)
Comment (bjhaid) [reply Documentation/gettingstarted/kubeproxy-free.rst]: Addressed in https://github.com/cilium/cilium/commit/262176f973869ede633241f226d038cf60a81de4

File: Documentation/gettingstarted/kubeproxy-free.rst (hunk 3 above)
Comment (bjhaid) [reply Documentation/gettingstarted/kubeproxy-free.rst]: Addressed in https://github.com/cilium/cilium/commit/262176f973869ede633241f226d038cf60a81de4

File: Documentation/gettingstarted/kubeproxy-free.rst (hunk 2 above)
Comment (bjhaid) [reply Documentation/gettingstarted/kubeproxy-free.rst]: Addressed in https://github.com/cilium/cilium/commit/262176f973869ede633241f226d038cf60a81de4

File: Documentation/gettingstarted/kubeproxy-free.rst [hunk 6]
```diff
@@ -1222,6 +1222,20 @@ working, take a look at `this KEP
...
     have the required label value.
 
+Neighbor Discovery
+******************
+
+When kube-proxy replacement is enabled Cilium does L2 neighbor discovery of nodes in the cluster.
+In some rare cases Cilium may leave stale entries behind in the neighbor table causing packets
+between some nodes to be dropped. To prevent Cilium from performing the neighbor discovery and
+instead rely on the Linux kernel to discover hosts on the same L2 network you can pass the
+`--enable-l2-neigh-discovery` flag to the cilium-agent setting it to false. Relying on the Linux
```
Comment (qmonnet) [inline Documentation/gettingstarted/kubeproxy-free.rst]:
```suggestion
``--enable-l2-neigh-discovery`` flag to the cilium-agent setting it to false. Relying on the Linux
```
We use RST markup, and we need double-backquotes instead of one like with Markdown.

**Review by qmonnet — CHANGES_REQUESTED** [review]

File: Documentation/gettingstarted/kubeproxy-free.rst (hunk 1 above)
Comment (qmonnet) [reply Documentation/gettingstarted/kubeproxy-free.rst]: It's nice to have this precision, but reading the whole paragraph, it sounds like the user is facing a choice between losing packets because of Cilium leaving stale entries _or_ because the kernel of a node has no neigh entry, so it's not really clear what's best.

Is there an advantage (like, a _lower_ chance to experience drops) when using the option? Or a recommendation such as “you may want this in _this_ case, but not in _that_ case”, that we could add to help the user understand? (@brb)

(Or maybe it's just me being picky - No strong objection on the current content of the note, I'm just wondering if we can make it clearer.)

File: Documentation/gettingstarted/kubeproxy-free.rst (hunk 6 above)
Comment (qmonnet) [reply Documentation/gettingstarted/kubeproxy-free.rst]: On the same sentence, I'd maybe rephrase to have `false` closer to the option name itself. The risk is that users might read the sentence to fast and stop after `you can pass the --enable-l2-neigh-discovery flag to the cilium-agent`.

**Review by brb — APPROVED** [review]
LGTM (minus what @qmonnet said). Thanks.

File: Documentation/gettingstarted/kubeproxy-free.rst (hunk 1 above)
Comment (bjhaid) [reply Documentation/gettingstarted/kubeproxy-free.rst]: >  it sounds like the user is facing a choice between losing packets because of Cilium leaving stale entries or because the kernel of a node has no neigh entry, so it's not really clear what's best.

In our setup and setups that use autoscaler this is true, to reduce the possibility of the neigh table entries getting gced, we already adjust `net.ipv4.neigh.default.gc_*` kernel params in our setup. Maybe adding details about setups using autoscaler or where nodes can go away and come back with the same name/ip but different mac address would be helpful... wdyt?

File: Documentation/gettingstarted/kubeproxy-free.rst (hunk 6 above)
Comment (bjhaid) [reply Documentation/gettingstarted/kubeproxy-free.rst]: @qmonnet can you propose a way through which I can rephrase the sentence?

Comment (qmonnet) [reply Documentation/gettingstarted/kubeproxy-free.rst]: How about:
```
To prevent [...] you can set the ``--enable-l2-neigh-discovery`` flag to ``false`` in the agent's configuration.
```
or
```
To prevent [...] you can pass ``--enable-l2-neigh-discovery=false`` to the cilium-agent.
```
or
```
To prevent [...], the cilium-agent has a ``--enable-l2-neigh-discovery`` flag that you can set to ``false``.
```
or just swapping the mention to the agent and the value:
```
To prevent [...] you can pass the ``--enable-l2-neigh-discovery`` flag, setting it to ``false``, to the cilium-agent.
```
Whichever sounds best to you.

Comment (bjhaid) [reply Documentation/gettingstarted/kubeproxy-free.rst]: addressed in https://github.com/cilium/cilium/commit/81553b9a88f7c2948666a721556895adb3bc6990.

Thanks!

File: Documentation/gettingstarted/kubeproxy-free.rst (hunk 1 above)
Comment (qmonnet) [reply Documentation/gettingstarted/kubeproxy-free.rst]: I don't know. This is maybe going too much into the details. Could you please just add an “also” to make it clearer that both cases may lead to some packet drops? Something like (emphasis on my additions):

“**However, note that** relying on the Linux Kernel might **also** cause some packets to be dropped ...”

And then let's leave it at this?

Comment (bjhaid) [reply Documentation/gettingstarted/kubeproxy-free.rst]: thanks @qmonnet , your recommendation is in https://github.com/cilium/cilium/commit/07d273ba20d93e9244af55bde920f0e7ef34861c

**Review by qmonnet — APPROVED** [review]

**Comment by qmonnet** [comment]
Checkpatch complains about the `Signed-off-by` tag, but it is present, just with a different name and address than the author registered from the git commit. Marking as ready to merge.

**Comment by bjhaid** [comment]
> Checkpatch complains about the Signed-off-by tag, but it is present, just with a different name and address than the author registered from the git commit. Marking as ready to merge.

Fixed commit author/email, thanks for catching it.
